src/analyzer/analyzerworker.cpp

```diff
@@ -0,0 +1,297 @@
+#include "analyzer/analyzerworker.h"
+
+#include <typeinfo>
+
+#include <QtDebug>
+#include <QThread>
+
+#ifdef __VAMP__
+#include "analyzer/analyzerbeats.h"
+#include "analyzer/analyzerkey.h"
+#include "analyzer/vamp/vampanalyzer.h"
+#endif
+#include "analyzer/analyzergain.h"
+#include "analyzer/analyzerebur128.h"
+#include "analyzer/analyzerwaveform.h"
+
+#include "sources/soundsourceproxy.h"
+#include "util/compatibility.h"
+#include "util/event.h"
+#include "util/timer.h"
+#include "util/trace.h"
+
+namespace {
+    // Analysis is done in blocks.
+    // We need to use a smaller block size, because on Linux the AnalyzerWorker
+    // can starve the CPU of its resources, resulting in xruns. A block size
+    // of 4096 frames per block seems to do fine.
+    const SINT kAnalysisChannels = mixxx::AudioSource::kChannelCountStereo;
+    const SINT kAnalysisFramesPerBlock = 4096;
+    const SINT kAnalysisSamplesPerBlock =
+        kAnalysisFramesPerBlock * kAnalysisChannels;
+} // anonymous namespace
+
+// Measured in 0.1%,
+// 0 for no progress during finalize
+// 1 to display the text "finalizing"
+// 100 for 10% step after finalize
+// NOTE: If this is changed, change woverview.cpp slotAnalyzerProgress().
+#define FINALIZE_PROMILLE 1
+
+// --- CONSTRUCTOR ---
+AnalyzerWorker::AnalyzerWorker(UserSettingsPointer pConfig, int workerIdx, bool priorized) :
+    m_pConfig(pConfig),
+    m_analyzelist(),
+    m_priorizedJob(priorized),
+    m_workerIdx(workerIdx),
+    m_sampleBuffer(kAnalysisSamplesPerBlock),
+    m_exit(false),
+    m_pauseRequested(false),
+    m_qm(),
+    m_qwait() {
+}
+
+// --- DESTRUCTOR ---
+AnalyzerWorker::~AnalyzerWorker() {
+    qDebug() << "Ending AnalyzerWorker";
+    // free resources
+    m_progressInfo.sema.release();
+ 
+    QListIterator<Analyzer*> it(m_analyzelist);
+    while (it.hasNext()) {
+        Analyzer* an = it.next();
+        delete an;
+    }
+}
+
+void AnalyzerWorker::nextTrack(TrackPointer newTrack) {
+    QMutexLocker locker(&m_qm);
+    m_currentTrack = newTrack;
```

Comment:
It would be an error to do resume on a worker that is waiting for a track. Also, I don't expect workers to be managed outside of AnalyzerManager, so basically the manager is the only responsible of interacting with them and is responsible of understanding their state.
calling EndProcess doesn't matter much, since it simply changes the exit flag. (it doesn't affect the thread state at that point).